From 8983bba90beb467d0346fada306fe94be4a5bdad Mon Sep 17 00:00:00 2001
From: "Rywacki, KamilX" <kamilx.rywacki@intel.com>
Date: Wed, 27 Mar 2019 17:07:23 +0100
Subject: [PATCH] PONRTSYS-3197: GSW_VlanFilterFree error during cleanup

- Clean VLAN Filters correctly in DP when we do not have any single or double tagged rules to set
- Assign "bEgressVlanFilter1Enable" and "bEgressVlanFilter2Enable" values correctly when we set Bridge Port configuration
---
 .../ethernet/lantiq/datapath/gswip31/datapath_tc_asym_vlan.c  | 11 +++++++++++
 drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c        |  4 ++--
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/datapath/gswip31/datapath_tc_asym_vlan.c b/drivers/net/ethernet/lantiq/datapath/gswip31/datapath_tc_asym_vlan.c
index 7c9cbb792210..2e0e41ac52ac 100644
--- a/drivers/net/ethernet/lantiq/datapath/gswip31/datapath_tc_asym_vlan.c
+++ b/drivers/net/ethernet/lantiq/datapath/gswip31/datapath_tc_asym_vlan.c
@@ -411,6 +411,17 @@ static int tc_vlan_filter(struct core_ops *ops,
 		j++;
 	}
 
+	/* Check if we have anything to configure */
+	if (j <= 0) {
+		/* Update bridge port */
+		ret = update_bp(ops,
+				(u32)info->bp,
+				vlan->dir == DP_DIR_INGRESS,
+				NULL,
+				NULL);
+		goto EXIT;
+	}
+
 	if (untagged < 0)
 		untagged = 0;
 	if (tagged < 0)
diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index 6de1479ed8a9..ec942a350d45 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -22981,7 +22981,7 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 			}
 		} else {
 			/* disable Egress VLAN filtering - Range 1*/
-			tbl_prog_brdgeport_ingress.val[10] &= ~(1 << 14);
+			tbl_prog_brdgeport_egress.val[10] &= ~(1 << 14);
 
 			if (gswdev->brdgeportconfig_idx[param->nBridgePortId].EgressVlanFilter1Assigned) {
 				idx = gswdev->brdgeportconfig_idx[param->nBridgePortId].EgressVlanFilter1BlkId;
@@ -23112,7 +23112,7 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 			}
 		} else {
 			/* disable Egress VLAN filtering - Range 2*/
-			tbl_prog_brdgeport_ingress.val[12] &= ~(1 << 14);
+			tbl_prog_brdgeport_egress.val[12] &= ~(1 << 14);
 
 			if (gswdev->brdgeportconfig_idx[param->nBridgePortId].EgressVlanFilter2Assigned) {
 				idx = gswdev->brdgeportconfig_idx[param->nBridgePortId].EgressVlanFilter2BlkId;
